Oxidizable and Waxable Blocks registries

Shnupbups opened the pull request:
Mods can add blocks that, like vanilla's Copper Blocks, can be waxed, stripped of wax, and/or stripped of oxidation decay.

I attempted to document it, let me know if it's sufficient.

This PR will be useful for mods attempting to add new Copper-based blocks into the game, such as Copper Buttons from the recent Mob Vote, without reimplementing a bunch of logic and/or using similar mixins to this PR.

Example:
```java
OxidizableBlocksRegistry.registerOxidizableBlockPair(ModBlocks.COPPER_BUTTON, ModBlocks.EXPOSED_COPPER_BUTTON);
OxidizableBlocksRegistry.registerOxidizableBlockPair(ModBlocks.EXPOSED_COPPER_BUTTON, ModBlocks.WEATHERED_COPPER_BUTTON);
OxidizableBlocksRegistry.registerOxidizableBlockPair(ModBlocks.WEATHERED_COPPER_BUTTON, ModBlocks.OXIDIZED_COPPER_BUTTON);

OxidizableBlocksRegistry.registerWaxableBlockPair(ModBlocks.COPPER_BUTTON, ModBlocks.WAXED_COPPER_BUTTON);
OxidizableBlocksRegistry.registerWaxableBlockPair(ModBlocks.EXPOSED_COPPER_BUTTON, ModBlocks.WAXED_EXPOSED_COPPER_BUTTON);
OxidizableBlocksRegistry.registerWaxableBlockPair(ModBlocks.WEATHERED_COPPER_BUTTON, ModBlocks.WAXED_WEATHERED_COPPER_BUTTON);
OxidizableBlocksRegistry.registerWaxableBlockPair(ModBlocks.OXIDIZED_COPPER_BUTTON, ModBlocks.WAXED_OXIDIZED_COPPER_BUTTON);
```

Shnupbups committed:
Oxidizable Blocks registry

Shnupbups committed:
Allow waxable blocks to be registered outside of OxidizableFamilies

Technici4n commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/OxidizableBlocksRegistry.java:
```diff
@@ -0,0 +1,213 @@
+	 * @param block the block whose oxidization is to be decreased
+	 * @return the block with decreased oxidization, in an optional
+	 */
+	public static Optional<Block> getDecreasedOxidizationBlock(Block block) {
```
Please use `@Nullable`, no `Optional`.

Shnupbups replied:
Optional is how the vanilla stuff does it, I'm sticking as close to the vanilla implementation as possible.

Technici4n commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/util/OxidizableFamily.java:
```diff
@@ -0,0 +1,192 @@
+ *
+ * <p>Each block also has a 'Waxed' form.
+ */
+public record OxidizableFamily(
```
Do you want this to have a public constructor? If not, I wouldn't use a record.

Shnupbups replied:
Having a public constructor is technically fine, though it does circumvent the null check that the Builder does. I suppose I can change it to a regular class.

liach replied:
you can write a canonical ctor that validates arguments, too

Technici4n commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/util/OxidizableFamily.java:
```diff
@@ -0,0 +1,192 @@
+		 * @return this builder
+		 */
+		public Builder add(OxidizationLevel level, Block unwaxed, Block waxed) {
+			if (!(unwaxed instanceof Oxidizable)) {
```
What is the point of this interface. 🤔

Shnupbups replied:
What interface? Oxidizable? That's a vanilla interface.

Technici4n commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/util/OxidizableFamily.java:
```diff
@@ -0,0 +1,192 @@
+		 */
+		public OxidizableFamily build() {
+			for (OxidizationLevel level : OxidizationLevel.values()) {
+				if (!unwaxed.containsKey(level) || unwaxed.get(level) == null) {
```
I'd just use `get() == null`.

Shnupbups replied:
I suppose that works.

Technici4n commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/OxidizableBlocksRegistry.java:
```diff
@@ -0,0 +1,213 @@
+	 * @param unwaxed the unwaxed variant
+	 * @param waxed the waxed variant
+	 */
+	public static void registerWaxable(Block unwaxed, Block waxed) {
```
This function and many others could use some `Objects.requireNonNull` assertions.

Shnupbups replied:
Good point, I never really use that myself but I should...

Technici4n commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/OxidizableBlocksRegistry.java:
```diff
@@ -0,0 +1,213 @@
+
+import net.fabricmc.fabric.api.util.OxidizableFamily;
+
+public class OxidizableBlocksRegistry {
```
I'd recommend letting people query the vanilla functions in `Oxidizable`. I'd only expose this API to inject into the relevant vanilla maps, which can be done by injecting into the synthetic lambda bodies: for example you can inject at the return of `method_34740` to replace the `ImmutableBiMap` by a regular `BiMap`, and later on append stuff to this `BiMap` directly.

Shnupbups replied:
The `BiMap` returned by `method_34740` is immediately memoized, so appending stuff to it later wouldn't do anything.

liach replied:
> The `BiMap` returned by `method_34740` is immediately memoized

Haha, reminds me of how mojang creates a whole map every time you want to query the vanilla furnace fuel values

Technici4n replied:
Yes, the reference to the map is memorized. But you can replace the immutable map by a regular map and that mutable map will be memoized. You can then edit it later by just querying the supplier.

Shnupbups replied:
Ah, yeah I see now. Alright, will do that then!

Shnupbups committed:
Refactor

Separated OxidizableBlocksRegistry and WaxableBlocksRegistry
Separated OxidizableFamily and WaxableBlockPair
Added null check to WaxableBlockPair
Added several methods

Shnupbups committed:
whoops lol

haykam821 commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/OxidizableBlocksRegistry.java:
```diff
@@ -52,31 +48,31 @@ public static void registerFamilies(OxidizableFamily... families) {
+	}
+
 	/**
 	 * Registers an {@link OxidizableFamily}s.
```
```suggestion
	 * Registers an {@link OxidizableFamily}.
```

haykam821 commented on fabric-content-registries-v0/src/testmod/java/net/fabricmc/fabric/test/content/registry/ContentRegistryTest.java:
```diff
@@ -39,11 +45,41 @@ public void onInitialize() {
 
 		TillableBlockRegistry.register(Blocks.GREEN_WOOL, context -> true, HoeItem.createTillAction(Blocks.LIME_WOOL.getDefaultState()));
+
+		// mote that since ores do not implement Oxidizable that a warning will be shown in the log
```
```suggestion
		// note that since ores do not implement Oxidizable that a warning will be shown in the log
```

Shnupbups committed:
Apply suggestions from code review

Co-authored-by: haykam821 <24855774+haykam821@users.noreply.github.com>

Shnupbups committed:
Update ContentRegistryTest.java

Shnupbups committed:
Refactor again lol

Now uses mixins to change the vanilla ImmutableBiMaps to HashBiMaps and adds to them instead of keeping separate maps and having methods of their own. Much cleaner!

Shnupbups committed:
gutted everything lol

removed all the charm and character- uh, i mean, superfluous stuff and reduced the api to the bare basics
oh and also one-way methods which is neat i guess if you're into that sort of thing

Shnupbups committed:
🦀one way methods are gone🦀

all my homies hate one way methods
maps are linked again

Shnupbups committed:
re-merged oxidizable and waxable registry classes

Shnupbups committed:
Update ContentRegistryTest.java

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/OxidizableBlocksRegistry.java:
```diff
@@ -0,0 +1,49 @@
+import net.minecraft.block.Oxidizable;
+import net.minecraft.item.HoneycombItem;
+
+public class OxidizableBlocksRegistry {
```
```suggestion
public final class OxidizableBlocksRegistry {
	private OxidizableBlocksRegistry() {
	}

```
To match other static method container classes

Juuxel replied:
This could do with a short (one-line) javadoc explaining that it's used for registering oxidisable and waxable blocks.

Juuxel commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/mixin/content/registry/OxidizableMixin.java:
```diff
@@ -0,0 +1,34 @@
+
+@Mixin(Oxidizable.class)
+public interface OxidizableMixin {
+	@Inject(method = "method_34740", at = @At("RETURN"), cancellable = true)
```
This is a lambda body, right? If so, a comment or a `Dynamic` would explain that random intermediary method for maintainers (this is also done in other modules)

Juuxel replied:
I meant more like explanation in the `Dynamic` like in https://github.com/FabricMC/fabric/blob/fa34ed2c10c399ed6c973a56f3fda31cd47811f6/fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/mixin/entity/event/LivingEntityMixin.java#L142

Juuxel reviewed (changes requested):
The API looks great, just some minor comments

Shnupbups committed:
implement suggestions by @Juuxel

Shnupbups committed:
Oxidization -> Oxidation

as per https://github.com/FabricMC/yarn/pull/2837

Shnupbups committed:
Improve Dynamics
